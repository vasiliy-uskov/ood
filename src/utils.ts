export function randomInt(min: number, max: number): number {
	return Math.floor(Math.random() * (max - min) + min)
}

export function generateUUIDv4(): string {
	let uuid = "";
	for (let i = 0; i < 16; ++i) {
		uuid = uuid.concat(randomInt(0, 255).toString(16))
	}
	return uuid;
}

export function generateDuplicatesBytes(byteValue: number, bytesCount: number): Array<number> {
	const bytes = [];
	for (let i = 0; i < bytesCount; ++i) {
		bytes.push(byteValue)
	}
	return bytes;
}