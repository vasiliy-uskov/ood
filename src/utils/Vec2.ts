class Vec2 {
	constructor(x: number, y: number) {
		this.x = x;
		this.y = y;
	}

	public static sum(...vectors: Array<Vec2>): Vec2 {
		return new Vec2(
			vectors.reduce((x, vec) => x + vec.x, 0),
			vectors.reduce((y, vec) => y + vec.y, 0)
		)
	}

	public static diff(a: Vec2, b: Vec2): Vec2 {
		return new Vec2(a.x - b.x, a.y - b.y)
	}

	public static scale(vec: Vec2, num: number): Vec2 {
		return new Vec2(vec.x * num, vec.y * num);
	}

	public static scalarComposite(a: Vec2, b: Vec2): number {
		return a.x * b.x + a.y * b.y
	}

	public normalize(): Vec2 {
		return (new PolarVec2(1, this.toPolar().angle)).toDescartes();
	}

	public reverse(): Vec2 {
		return Vec2.scale(this, -1);
	}

	public toPolar(): PolarVec2 {
		if (!this._polarVec2)
		{
			const radius = Math.sqrt(this.x * this.x + this.y * this.y);
			const polarBasis = new Vec2(1, 0);
			const cos = Vec2.scalarComposite(this, polarBasis) / radius;
			const angle = Math.sign(this.y) * Math.acos(cos);
			this._polarVec2 = new PolarVec2(radius, angle);
		}
		return this._polarVec2;
	}

	public readonly x: number;
	public readonly y: number;

	private _polarVec2: PolarVec2 | null = null;
}

class PolarVec2 {
	constructor(radius: number, angle: number) {
		this.radius = radius;
		this.angle = angle;
	}

	public toDescartes(): Vec2 {
		if (!this._descartes)
		{
			this._descartes = new Vec2(
				this.radius * Math.cos(this.angle),
				this.radius * Math.sin(this.angle)
			);
		}
		return this._descartes;
	}

	public readonly angle: number;
	public readonly radius: number;

	private _descartes: Vec2 | null = null;
}

export {
	Vec2,
};