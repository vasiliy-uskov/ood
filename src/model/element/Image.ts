import {Size} from "../../utils/Size";
import {IReadonlyResource, IResource} from "../resource/IResource";

export class Image {
	public readonly size: Size;
	public readonly img: IResource;
}

export class ReadonlyImage {
	public readonly size: Size;
	public readonly img: IReadonlyResource;
}