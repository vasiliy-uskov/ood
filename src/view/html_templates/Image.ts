import {Size} from "../../utils/Size";

export type ImageProps = {
	readonly src: string,
	readonly size: Size
};

export function Image(props: ImageProps) {
	return `<img src="${props.src}" width="${props.size.width}" height="${props.size.height}" alt="" />`;
}