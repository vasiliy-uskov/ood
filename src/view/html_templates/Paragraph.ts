import {escapeHtmlText} from "../../utils/utils";

export type ParagraphProps = {
	readonly text: string
};

export function Paragraph(props: ParagraphProps) {
	return `<p>${escapeHtmlText(props.text)}</p>`
}