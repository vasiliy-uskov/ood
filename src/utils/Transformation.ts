import {Vec2} from "./Vec2";

type TransformConfig = {
	_00: number, _01: number, _02: number,
	_10: number, _11: number, _12: number,
}

class Transformation {
	constructor(config: TransformConfig) {
		this._points = config;
	}

	public add(tansform: Transformation): Transformation {
		const points1 = Object.assign(tansform._points);
		const points2 = Object.assign(this._points);
		points1._00 = points1._00 * points2._00 + points1._01 * points2._10;
		points1._01 = points1._00 * points2._01 + points1._01 * points2._11;
		points1._10 = points1._10 * points2._00 + points1._11 * points2._10;
		points1._11 = points1._10 * points2._01 + points1._11 * points2._11;
		points1._02 = points1._00 * points2._02 + points1._01 * points2._12 + points1._02;
		points1._12 = points1._10 * points2._02 + points1._11 * points2._12 + points1._12;
		this._points = points1;
		return this;
	}

	public rotate(angle: number): Transformation {
		return this.add(Transformation.rotate(angle));
	}

	public scale(scaleX: number, scaleY: number): Transformation {
		return this.add(Transformation.scale(scaleX, scaleY));
	}

	public translate(vec: Vec2): Transformation {
		return this.add(Transformation.translate(vec));
	}

	public useTransformOrigin(origin: Vec2): Transformation {
		return Transformation
			.translate(origin.reverse())
			.add(this)
			.translate(origin);
	}

	public transform(vec: Vec2): Vec2 {
		const x = vec.x * this._points._00 + vec.y * this._points._01 + this._points._02;
		const y = vec.x * this._points._10 + vec.y * this._points._11 + this._points._12;
		return new Vec2(x, y);
	}

	public static rotate(angle: number): Transformation {
		return new Transformation({
			_00: Math.cos(-angle), _01: Math.sin(-angle), _02: 0,
			_10: -Math.sin(-angle), _11: Math.cos(-angle), _12: 0,
		})
	}

	public static scale(scaleX: number, scaleY: number = scaleX): Transformation {
		return new Transformation({
			_00: scaleX, _01: 0, _02: 0,
			_10: 0, _11: scaleY, _12: 0,
		})
	}

	public static translate(vec: Vec2): Transformation {
		return new Transformation({
			_00: 1, _01: 0, _02: vec.x,
			_10: 0, _11: 1, _12: vec.y,
		})
	}

	public static empty(): Transformation {
		return new Transformation({
			_00: 1, _01: 0, _02: 0,
			_10: 0, _11: 1, _12: 0,
		})
	}

	private _points: TransformConfig;
}

export {Transformation};