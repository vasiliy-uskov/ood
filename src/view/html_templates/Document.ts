import {escapeHtmlText} from "../../utils/utils";
import {Image, ImageProps} from "./Image";
import {Paragraph, ParagraphProps} from "./Paragraph";

export enum ContentType {
	PARAGRAPH = 'paragraph',
	IMAGE = 'image',
}

export type ParagraphDocumentItem = {
	type: ContentType.PARAGRAPH,
	props: ParagraphProps,
}

export type ImageDocumentItem = {
	type: ContentType.IMAGE,
	props: ImageProps,
}

export type DocumentContent = ReadonlyArray<(ParagraphDocumentItem|ImageDocumentItem)>;

export type DocumentProps = {
	title: string,
	content: DocumentContent,
}

export function Document(props: DocumentProps) {
	return `<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>${escapeHtmlText(props.title)}</title>
</head>
<body>
	${props.content.map(item => {
		switch (item.type) {
			case ContentType.IMAGE:
				return Image(item.props);
			case ContentType.PARAGRAPH:
				return Paragraph(item.props);
			default:
				throw new Error("Undefined content type")
		}
	}).join("")}
</body>
</html>`;
}