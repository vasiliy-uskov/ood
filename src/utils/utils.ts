import {Vec2} from "./Vec2";
import {Transformation} from "./Transformation";

export function isDefAndNotNull<T>(x: T | null | undefined): x is T {
	return x !== null && x !== undefined;
}

export function getDotsByBasis(leftTop: Vec2, widthBasis: Vec2, heightBasis: Vec2): [Vec2, Vec2, Vec2, Vec2] {
	return [
		leftTop,
		Vec2.sum(leftTop, widthBasis),
		Vec2.sum(leftTop, widthBasis, heightBasis),
		Vec2.sum(leftTop, heightBasis),
	]
}

type transformBasisProps = Readonly<{
	transformation: Transformation
	transformOrigin: Vec2,
	center: Vec2,
	basisX: Vec2,
	basisY: Vec2
}>

export function transformBasis({transformation, transformOrigin, center, basisX, basisY}: transformBasisProps): [Vec2, Vec2, Vec2] {
	const transformationWithOrigin = transformation.useTransformOrigin(transformOrigin);
	const newCenter = transformationWithOrigin.transform(center);
	const newBasisX = Vec2.diff(transformationWithOrigin.transform(Vec2.sum(basisX, center)), newCenter);
	const newBasisY = Vec2.diff(transformationWithOrigin.transform(Vec2.sum(basisY, center)), newCenter);
	return [newCenter, newBasisX, newBasisY];
}