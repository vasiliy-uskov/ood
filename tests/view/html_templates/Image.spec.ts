import {Image} from "../../../src/view/html_templates/Image";
import {Size} from "../../../src/utils/Size";

function expandImage(src: string, size: Size) {
	expect(Image({src, size})).toMatchSnapshot();
}

it('expand image', () => {
	expandImage('./location/img.png', {
		width: 100,
		height: 100,
	});
});