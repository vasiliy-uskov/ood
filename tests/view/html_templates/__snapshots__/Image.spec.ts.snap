// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expand image 1`] = `"<img src=\\"./location/img.png\\" width=\\"100\\" height=\\"100\\" alt=\\"\\" />"`;
